--- origsrc/mDNSResponder-625.41.2/mDNSShared/CommonServices.h	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSShared/CommonServices.h	2016-09-08 18:28:28.669290600 -0500
@@ -260,6 +260,10 @@ extern "C" {
         #pragma warning( disable:4127 ) // Disable "conditional expression is constant" warning for debug macros.
         #pragma warning( disable:4706 ) // Disable "assignment within conditional expression" for Microsoft headers.
 
+    #elif ( defined( __MINGW32__ ) )
+
+        #include    <stdint.h>
+
     #endif
 
     #include    <windows.h>
@@ -474,7 +478,7 @@ typedef int socklen_t;
 // - Windows
 
 #if ( TARGET_LANGUAGE_C_LIKE )
-    #if ( !defined(_SSIZE_T) && ( TARGET_OS_WIN32 || !defined( _BSD_SSIZE_T_DEFINED_ ) ) && !TARGET_OS_FREEBSD && !TARGET_OS_LINUX && !TARGET_OS_VXWORKS && !TARGET_OS_MAC)
+    #if ( !defined(_SSIZE_T) && ( TARGET_OS_WIN32 || !defined( _BSD_SSIZE_T_DEFINED_ ) ) && !TARGET_OS_FREEBSD && !TARGET_OS_LINUX && !TARGET_OS_VXWORKS && !TARGET_OS_MAC && !defined(__MINGW32__))
 typedef int ssize_t;
     #endif
 #endif
--- origsrc/mDNSResponder-625.41.2/mDNSShared/uds_daemon.c	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSShared/uds_daemon.c	2016-09-08 18:28:28.678290000 -0500
@@ -17,6 +17,7 @@
 
 #if defined(_WIN32)
 #include <process.h>
+#include <ws2tcpip.h>
 #define usleep(X) Sleep(((X)+999)/1000)
 #else
 #include <fcntl.h>
@@ -134,7 +135,7 @@ typedef struct browser_t
     DNSQuestion q;
 } browser_t;
 
-#ifdef _WIN32
+#ifdef _MSC_VER
     typedef unsigned int pid_t;
     typedef unsigned int socklen_t;
 #endif
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/DLL/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/DLL/Makefile	2016-09-08 19:02:14.062669000 -0500
@@ -0,0 +1,59 @@
+OBJDIR = objects/prod
+BUILDDIR = build/prod
+COREDIR = ../../mDNSCore
+SHAREDDIR = ../../mDNSShared
+MACOSXDIR = ../../mDNSMacOSX
+WINDOWSDIR = ..
+
+ARCH = `arch`
+CC = $(ARCH)-w64-mingw32-gcc
+RC = $(ARCH)-w64-mingw32-windres
+CFLAGS_COMMON = -I$(COREDIR) -I$(WINDOWSDIR) -I$(SHAREDDIR)
+CFLAGS_DEBUG = -O2 -g
+CFLAGS_OS = -DWIN32 -DNDEBUG -D_WINDOWS -D_USRDLL -DMDNS_DEBUGMSGS=0 -DWIN32_LEAN_AND_MEAN -DUSE_TCP_LOOPBACK -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES=1 -DNOT_HAVE_SA_LEN
+CFLAGS = $(CFLAGS_COMMON) $(CFLAGS_OS) $(CFLAGS_DEBUG)
+LIBS = -lws2_32 -liphlpapi
+
+OBJS = 					\
+	$(OBJDIR)/DebugServices.c.o	\
+	$(OBJDIR)/dll.rc.o		\
+	$(OBJDIR)/dllmain.c.o		\
+	$(OBJDIR)/dnssd_clientlib.c.o	\
+	$(OBJDIR)/dnssd_clientstub.c.o	\
+	$(OBJDIR)/dnssd_ipc.c.o		\
+	$(OBJDIR)/GenLinkedList.c.o
+
+all: setup libdns_sd
+
+# 'setup' sets up the build directory structure the way we want
+setup:
+	@if test ! -d $(OBJDIR)   ; then mkdir -p $(OBJDIR)   ; fi
+	@if test ! -d $(BUILDDIR) ; then mkdir -p $(BUILDDIR) ; fi
+
+libdns_sd: setup $(BUILDDIR)/dnssd.dll
+	@echo "Client library done"
+
+$(BUILDDIR)/dnssd.dll: $(OBJS) dnssd.def
+	$(CC) -shared -Wl,--out-implib,$(BUILDDIR)/libdns_sd.dll.a -o $@ $^ $(LIBS)
+
+clean:
+	rm -f $(BUILDDIR)/dnssd.dll $(OBJS)
+
+# Implicit rules
+$(OBJDIR)/%.c.o:	%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(COREDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(SHAREDDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(MACOSXDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(WINDOWSDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.rc.o:	%.rc
+	$(RC) -I$(WINDOWSDIR) -o $@ $<
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/SystemService/EventLog.h	1969-12-31 18:00:00.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/SystemService/EventLog.h	2016-09-08 18:28:28.682290900 -0500
@@ -0,0 +1 @@
+#define MDNSRESPONDER_LOG ((WORD)100)
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/SystemService/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/SystemService/Makefile	2016-09-08 19:01:59.026490300 -0500
@@ -0,0 +1,70 @@
+OBJDIR = objects/prod
+BUILDDIR = build/prod
+COREDIR = ../../mDNSCore
+SHAREDDIR = ../../mDNSShared
+MACOSXDIR = ../../mDNSMacOSX
+WINDOWSDIR = ..
+
+ARCH = `arch`
+CC = $(ARCH)-w64-mingw32-gcc -municode
+RC = $(ARCH)-w64-mingw32-windres
+CFLAGS_COMMON = -I$(COREDIR) -I$(WINDOWSDIR) -I$(SHAREDDIR)
+CFLAGS_DEBUG = -O2 -g
+CFLAGS_OS = -DWIN32 -DNDEBUG -D_CONSOLE -DTARGET_OS_WIN32 -DWIN32_LEAN_AND_MEAN -DUSE_TCP_LOOPBACK -DPLATFORM_NO_STRSEP -DPLATFORM_NO_EPIPE -DPLATFORM_NO_RLIMIT -DPID_FILE=\"\" -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES=1 -D_LEGACY_NAT_TRAVERSAL_
+CFLAGS = $(CFLAGS_COMMON) $(CFLAGS_OS) $(CFLAGS_DEBUG)
+LIBS = -lws2_32 -liphlpapi -lnetapi32 -lpowrprof
+
+OBJS = 						\
+	$(OBJDIR)/anonymous.c.o			\
+	$(OBJDIR)/CryptoAlg.c.o			\
+	$(OBJDIR)/DebugServices.c.o		\
+	$(OBJDIR)/DNSCommon.c.o			\
+	$(OBJDIR)/DNSDigest.c.o			\
+	$(OBJDIR)/dnssd_ipc.c.o			\
+	$(OBJDIR)/GenLinkedList.c.o		\
+	$(OBJDIR)/LegacyNATTraversal.c.o	\
+	$(OBJDIR)/main.c.o			\
+	$(OBJDIR)/mDNS.c.o			\
+	$(OBJDIR)/mDNSDebug.c.o			\
+	$(OBJDIR)/mDNSWin32.c.o			\
+	$(OBJDIR)/Poll.c.o			\
+	$(OBJDIR)/Secret.c.o			\
+	$(OBJDIR)/Service.c.o			\
+	$(OBJDIR)/Service.rc.o			\
+	$(OBJDIR)/uDNS.c.o			\
+	$(OBJDIR)/uds_daemon.c.o
+
+all: setup Daemon
+
+# 'setup' sets up the build directory structure the way we want
+setup:
+	@if test ! -d $(OBJDIR)   ; then mkdir -p $(OBJDIR)   ; fi
+	@if test ! -d $(BUILDDIR) ; then mkdir -p $(BUILDDIR) ; fi
+
+Daemon: setup $(BUILDDIR)/mDNSResponder.exe
+	@echo "Responder daemon done"
+
+$(BUILDDIR)/mDNSResponder.exe: $(OBJS)
+	$(CC) -static -o $@ $^ $(LIBS)
+
+clean:
+	rm -f $(BUILDDIR)/mDNSResponder.exe $(OBJS)
+
+# Implicit rules
+$(OBJDIR)/%.c.o:	%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(COREDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(SHAREDDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(MACOSXDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(WINDOWSDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.rc.o:	%.rc
+	$(RC) -I$(WINDOWSDIR) -o $@ $<
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/SystemService/Service.c	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/SystemService/Service.c	2016-09-08 18:28:28.693291700 -0500
@@ -38,7 +38,9 @@
 #include	"mDNSWin32.h"
 #include	"mDNSDebug.h"
 
+#ifndef __MINGW32__
 #include	"Firewall.h"
+#endif
 
 #if( !TARGET_OS_WINDOWS_CE )
 	#include	<mswsock.h>
@@ -665,8 +667,10 @@ static OSStatus CheckFirewall()
 		err = translate_errno( size > 0, (OSStatus) GetLastError(), kPathErr );
 		require_noerr( err, exit );
 
+#ifndef __MINGW32__
 		err = mDNSAddToFirewall(fullPath, kServiceFirewallName);
 		require_noerr( err, exit );
+#endif
 
 		value = 1;
 		err = RegSetValueEx( key, kServiceManageFirewall, 0, REG_DWORD, (const LPBYTE) &value, sizeof( DWORD ) );
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/SystemService/Service.rc	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/SystemService/Service.rc	2016-09-08 18:54:42.519863700 -0500
@@ -7,9 +7,11 @@
 //
 // Generated from the TEXTINCLUDE 2 resource.
 //
-#include "winres.h"
+#include "afxres.h"
 #include "WinVersRes.h"
+#if 0
 #include "EventLog.rc"
+#endif
 /////////////////////////////////////////////////////////////////////////////
 #undef APSTUDIO_READONLY_SYMBOLS
 
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/mDNSWin32.c	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/mDNSWin32.c	2016-09-08 18:28:28.703303600 -0500
@@ -35,7 +35,9 @@
 #include	"Poll.h"
 #include	"CommonServices.h"
 #include	"DebugServices.h"
+#ifndef __MINGW32__
 #include	"Firewall.h"
+#endif
 #include	"RegNames.h"
 #include	"Secret.h"
 #include	<dns_sd.h>
@@ -4932,7 +4934,11 @@ CheckFileShares( mDNS * const m )
 		RegQueryValueEx( key, L"Advertise", NULL, NULL, (LPBYTE) &advertise, &dwSize );
 	}
 
+#ifndef __MINGW32__
 	if ( advertise && mDNSIsFileAndPrintSharingEnabled( &retry ) )
+#else
+	if ( advertise )
+#endif
 	{
 		dlog( kDebugLevelTrace, DEBUG_NAME "Sharing is enabled\n" );
 
