--- origsrc/mDNSResponder-625.41.2/mDNSShared/CommonServices.h	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSShared/CommonServices.h	2016-09-08 18:28:28.669290600 -0500
@@ -260,11 +260,15 @@ extern "C" {
         #pragma warning( disable:4127 ) // Disable "conditional expression is constant" warning for debug macros.
         #pragma warning( disable:4706 ) // Disable "assignment within conditional expression" for Microsoft headers.
 
+    #elif ( defined( __MINGW32__ ) )
+
+        #include    <stdint.h>
+
     #endif
 
     #include    <windows.h>
     #include    <winsock2.h>
-    #include    <Ws2tcpip.h>
+    #include    <ws2tcpip.h>
 
     #if ( defined( _MSC_VER ) )
         #pragma warning( default:4706 )
@@ -474,7 +478,7 @@ typedef int socklen_t;
 // - Windows
 
 #if ( TARGET_LANGUAGE_C_LIKE )
-    #if ( !defined(_SSIZE_T) && ( TARGET_OS_WIN32 || !defined( _BSD_SSIZE_T_DEFINED_ ) ) && !TARGET_OS_FREEBSD && !TARGET_OS_LINUX && !TARGET_OS_VXWORKS && !TARGET_OS_MAC)
+    #if ( !defined(_SSIZE_T) && ( TARGET_OS_WIN32 || !defined( _BSD_SSIZE_T_DEFINED_ ) ) && !TARGET_OS_FREEBSD && !TARGET_OS_LINUX && !TARGET_OS_VXWORKS && !TARGET_OS_MAC && !defined(__MINGW32__))
 typedef int ssize_t;
     #endif
 #endif
--- origsrc/mDNSResponder-625.41.2/mDNSShared/uds_daemon.c	2016-03-11 03:47:07.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSShared/uds_daemon.c	2016-09-08 18:28:28.678290000 -0500
@@ -17,6 +17,7 @@
 
 #if defined(_WIN32)
 #include <process.h>
+#include <ws2tcpip.h>
 #define usleep(X) Sleep(((X)+999)/1000)
 #else
 #include <fcntl.h>
@@ -134,7 +135,7 @@ typedef struct browser_t
     DNSQuestion q;
 } browser_t;
 
-#ifdef _WIN32
+#ifdef _MSC_VER
     typedef unsigned int pid_t;
     typedef unsigned int socklen_t;
 #endif
--- origsrc/mDNSResponder-765.50.9/mDNSWindows/mDNSWin32.c	2016-09-13 17:37:18.000000000 -0500
+++ src/mDNSResponder-765.50.9/mDNSWindows/mDNSWin32.c	2017-09-13 11:53:30.246082900 -0500
@@ -40,7 +40,7 @@
 #include	"Secret.h"
 #include	<dns_sd.h>
 
-#include	<Iphlpapi.h>
+#include	<iphlpapi.h>
 #include	<mswsock.h>
 #include	<process.h>
 #include	<ntsecapi.h>
--- origsrc/mDNSResponder-625.41.2/mDNSWindows/DLL/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ src/mDNSResponder-625.41.2/mDNSWindows/DLL/Makefile	2016-09-08 19:02:14.062669000 -0500
@@ -0,0 +1,59 @@
+OBJDIR = objects/prod
+BUILDDIR = build/prod
+COREDIR = ../../mDNSCore
+SHAREDDIR = ../../mDNSShared
+MACOSXDIR = ../../mDNSMacOSX
+WINDOWSDIR = ..
+
+ARCH = `arch`
+CC = $(ARCH)-w64-mingw32-gcc
+RC = $(ARCH)-w64-mingw32-windres
+CFLAGS_COMMON = -I$(COREDIR) -I$(WINDOWSDIR) -I$(SHAREDDIR)
+CFLAGS_DEBUG = -O2 -g
+CFLAGS_OS = -DWIN32 -DNDEBUG -D_WINDOWS -D_USRDLL -DMDNS_DEBUGMSGS=0 -DWIN32_LEAN_AND_MEAN -DUSE_TCP_LOOPBACK -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES=1 -DNOT_HAVE_SA_LEN
+CFLAGS = $(CFLAGS_COMMON) $(CFLAGS_OS) $(CFLAGS_DEBUG)
+LIBS = -lws2_32 -liphlpapi
+
+OBJS = 					\
+	$(OBJDIR)/DebugServices.c.o	\
+	$(OBJDIR)/dll.rc.o		\
+	$(OBJDIR)/dllmain.c.o		\
+	$(OBJDIR)/dnssd_clientlib.c.o	\
+	$(OBJDIR)/dnssd_clientstub.c.o	\
+	$(OBJDIR)/dnssd_ipc.c.o		\
+	$(OBJDIR)/GenLinkedList.c.o
+
+all: setup libdns_sd
+
+# 'setup' sets up the build directory structure the way we want
+setup:
+	@if test ! -d $(OBJDIR)   ; then mkdir -p $(OBJDIR)   ; fi
+	@if test ! -d $(BUILDDIR) ; then mkdir -p $(BUILDDIR) ; fi
+
+libdns_sd: setup $(BUILDDIR)/dnssd.dll
+	@echo "Client library done"
+
+$(BUILDDIR)/dnssd.dll: $(OBJS) dnssd.def
+	$(CC) -shared -Wl,--out-implib,$(BUILDDIR)/libdns_sd.dll.a -o $@ $^ $(LIBS)
+
+clean:
+	rm -f $(BUILDDIR)/dnssd.dll $(OBJS)
+
+# Implicit rules
+$(OBJDIR)/%.c.o:	%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(COREDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(SHAREDDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(MACOSXDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.c.o:	$(WINDOWSDIR)/%.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(OBJDIR)/%.rc.o:	%.rc
+	$(RC) -I$(WINDOWSDIR) -o $@ $<
